libraries/standard/mqtt/src/mqtt.c

```diff
@@ -375,41 +438,41 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * const pContext,
 
     if( status == MQTTSuccess )
     {
-        /* Send header first. */
-        bytesSent = sendPacket( pContext,
-                                pContext->networkBuffer.pBuffer,
-                                headerSize );
-
-        if( bytesSent < 0 )
+        /* Reserve state for publish message. Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
```

Comment:
This isn't strictly necessary since `MQTT_ReserveState()` was written with the possibility of calling it with QoS 0 in mind. However, since that function is meant to be internal use only, it presents us with the opportunity to convert its checks into asserts and reduce the amount of code. It isn't necessary for this PR, but we need to remember to go back to it and remove its unnecessary code.

Reply:
I don't totally agree with this. Our current comments (https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/libraries/standard/mqtt/include/mqtt_state.h#L48) indicates that it is either QoS1 or 2. Also I think if it has to be called only for QoS1 or 2, it will definitely help the readability of the function that the state machine has to be invoked only specific QoSs instead for all and then asserting it. And what happens for QoS0 publishes with asserts enabled.

Reply:
There might be some confusion, but my suggestion was to update `MQTT_ReserveState()` to change its `if(qos == MQTTQoS0)` into just an `assert(qos != MQTTQoS0)`, since this should be the only place that function is called from. The code in this file should stay as it is.

Reply:
Ok. Yea I was thinking to update the call sites for the function. I have again a question related to the proposed optimization regarding the function, Isn't MQTT_ReserveState() a public API function? Asserts are not used in a public API. Given the composable tenet that we are following, I think it makes sense for us to have the state machine available for the customers in its full form rather than making MQTT_ReserveState() an internal function. What do you think?
Anyway, I think this can be handled in a separate PR based on what we decide.